### ISSUE TITLE
Azure SDK for Java 1.0.0

### ISSUE THREAD

**Issue by jianghaolu** [Azure/azure-sdk-for-java#568]
This is a simple living spec of Azure SDK for Java 1.0.0. Comments are welcome.

## Goals
- Azure SDK for Java 0.9.x and previous versions ("old SDK") are generated from an internal Web API specs through an internal generator, which are both deprecated. Starting from Azure SDK for Java 1.0 ("new SDK") clients are generated from [Swagger 2.0 specs](https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md) with [AutoRest](https://github.com/azure/autorest) client generator.
- The old SDK was first written years ago and many libraries and practices used are outdated. the new SDK aims to adapt more modern and open technologies.
- Using the old SDK can be painful in constructing a client, dealing with authentication, and making asynchronous calls, etc. The new SDK aims to improve much in usability.
- The old SDK doesn't support Android and the new one aims to run on Android effortlessly. There was an old Android SDK but has been quite outdated too. By supporting Android, the new SDK is not just "able to run" on Android - many have been designed with the consideration of Android development scenarios, e.g. callback based asynchronous calls.

## Repository Structure
The project structure is shown below:
```
 .
├── azure/
│   ├── pom.xml
│   └── src/config/checkstyle/
├── azure-mgmt-{service}/
│   ├── pom.xml
│   ├── src/
│   │   ├── main/java/com/microsoft/azure/management/{service}/
│   │   │   └── models/
│   │   └── test/java/com/microsoft/azure/management/{service}/
├── tools/
└── pom.xml
```
`{service}` means any Azure Resource Manager service. We've currently generated compute, storage, resources, and network packages. `azure/pom.xml` is still the parent pom.

### Runtime
The one directory that is missing in the new SDK is the `core/` folder which contained the runtime for the old SDK. The runtime for the new SDK has been moved to https://github.com/Azure/autorest/tree/dev/ClientRuntimes/Java. There are 4 packages for runtime. Azure-client-runtime together with client-runtime serves the purpose of `core/` in 0.9 SDK. The other two authentication packages are added in 1.0 as ADAL connectors for authentication purposes.
- client-runtime *basic functionality for validation, serialization, exception handling, authentication, etc*
- azure-client-runtime *Azure specific functionality for asynchronous operations, paging operations, etc*
- azure-client-authentication *OrgId and service principal authentication functionalities through ADAL on JDK*
- azure-android-client-authentication *OrgId and service principal authentication functionalities through ADAL on Android*

Client-runtime, azure-client-runtime and SDK code are designed to run on both JDK 1.7+ and Android 4.0+. Azure-client-authentication runs on JDK 1.7+ and azure-android-client-authentication runs on Android 4.0+.

### REST client
The new SDK for Java delegates the REST layer to [Retrofit](http://square.github.io/retrofit/) designed by Square engineers, which is a REST client designed with ease of use and great extensibility on JDK and Android. Declaring a REST API is simple with Retrofit,
```java
public interface Service {
  @GET("/Microsoft.Resources/{resourceGroupName}/list")
  Call<ResponseBody> list(@Path("resourceGroupName") String resourceGroupName);
}
```
SDK will handle the `ResponseBody` returned and deserialize into correctly output types.

### HTTP client
We use the default HTTP client by Retrofit, [OkHttp](http://square.github.io/okhttp/). OkHttp is lightweight and easy to use with Retrofit.

### JSON Serializer
We use [Jackson](http://wiki.fasterxml.com/JacksonHome) for serialization in the new SDK. Since the new SDK is only for Azure Resource Manager, there is no XML serialization involved.

### Client Initialization
Service loader is no longer required for creating a management client. You can simply call the constructor on a management client's implementation.
```java
ApplicationTokenCredentials credentials = new ApplicationTokenCredentials(
        "client-id", "domain", "client-secret", null);
resourceManagementClient = new ResourceManagementClientImpl(credentials);
resourceManagementClient.setSubscriptionId("subscription-GUID");
resourceManagementClient.setLogLevel(HttpLoggingInterceptor.Level.BODY);
```

### Synchronous Calls
A sample:
```java
ServiceResponse<List<Pet>> response = client.getPet().list();
List<Pet> response = client.getPet().List().getBody();
```
`ServiceResponse` contains the information of the request, response, status code, and the response body.

### Asynchronous Calls
A sample:
```java
Call<ResponseBody> call = client.getPet().list(new ServiceCallback<List<Pet>>) {
    @Override
    public void failure(Throwable t) { }

	@Override
	public void success(ServiceResponse<List<Pet>> response) {
		runOnUiThread(new Runnable() {
		    public void run() {
		    	updatePets(response.getBody());
		    }
		});
	}
});
...
call.cancel();
```

### Paging Operation
Paging operations are operations that return a big list of resources, split into multiple pages.
```java
Page<Pet> pets = client.list().getBody(); // fetches the first page
addPetsToDatabase(pets.getItems());
while (pets.getNextPageLink() != null) {
    pets = client.listNext(pets.getNextPageLink()).getBody(); // subsequent pages
    addPetsToDatabase(pets.getItems());
}
```

### Packages
| Scenario | 0.9 | 1.0 |
|--------- | --- | --- |
| runtime | com.microsoft.windowsazure(.core) | com.microsoft.rest |
| ARM code | com.microsoft.azure.compute | com.microsoft.azure.compute |

### Client
| Scenario | 0.9 | 1.0 |
|--------- | --- | --- |
| interface | ComputeManagementClient | ComputeManagementClient |
| implementation | ComputeManagementClientImpl | ComputeManagementClientImpl |
The names in 1.0 are generated from the `title` info in the Swagger spec:
```json
{
  "swagger": "2.0",
  "info": {
    "title": "ComputeManagementClient",
```

### Method groups (Operations)
| Scenario | 0.9 | 1.0 |
|--------- | --- | --- |
| Class name | VirtualMachineOperations | VirtualMachineOperations |
| getter on client | getVirtualMachineOperations() | getVirtualMachines() |
The names in 1.0 are generated from the `operationId`'s prefix in the Swagger spec:
```json
"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/virtualMachines/{vmName}/capture": {
  "post": {
    "operationId": "VirtualMachines_Capture",
```
If the method group name is a reserved word in Java, "Operations" is appended.
### Methods
| Scenario | 0.9 | 1.0 |
|--------- | --- | --- |
| method name | capture() | capture() |
The names in 1.0 are generated from the `operationId` in the Swagger spec:
```json
"/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Compute/virtualMachines/{vmName}/capture": {
  "post": {
    "operationId": "VirtualMachines_Capture",
```
If the method name is a reserved word in Java, "Method" is appended.

### Base Resource Models
Models extend from the following classes:

| Scenario | 0.9 | 1.0 |
|--------- | --- | --- |
| Empty Resource | N/A | BaseResource |
| Resource with location and tags | BaseResource | N/A |
| Resource with id, name, type, tags, location | BaseResourceExtended | Resource |

### Models
| Scenario | 0.9 | 1.0 |
|--------- | --- | --- |
| Class definition | `VirtualMachine extends ResourceBaseExtended` | `VirtualMachine extends Resource` |
To define a resource in Swagger spec for 1.0, 
```json
"AvailabilitySet": {
  "allOf": [
    {
      "$ref": "#/definitions/Resource"
    }
  ],
```

### Responses
| Scenario | 0.9 | 1.0 |
|--------- | --- | --- |
| Normal response | `VirtualMachineSizeListResponse extends OperationResponse` | `VirtualMachineSizeListResult` |
| Paging response | VirtualMachineListResponse (contains `getNextLink()`) | ServiceResponse&lt;Page&lt;VirtualMachine&gt;&gt; (`getNextLink()` is a method of `Page<?>`) |
| Response w/o body | DeleteOperationResponse | ServiceResponse&lt;Void&gt; |
| HEAD operation response | N/A | ServiceResponse&lt;Boolean&gt; |

### Exceptions
| Scenario | 0.9 | 1.0 |
|--------- | --- | --- |
| Base exception type | N/A | AutoRestException |
| Default exception type | ServiceException | CloudException |

### Credentials
User certificate credential has been removed in 1.0. User certificate authentication is only used for Service management. Service principal certificate authentication may be supported in a future version.
For token credentials:

| Scenario | 0.9 | 1.0 |
|--------- | --- | --- |
| Base type | TokenCloudCredentials | TokenCredentials |
| Implementations of token acquisition | N/A | UserTokenCredentials, ApplicationTokenCredentials |

### 0.9
```java
// Create a configuration object
Configuration config = ManagementConfiguration.configure(
        null,
        "localhost:3000",
        "subscription id in GUID",
        getAccessTokenFromUserCredentials("username", "Pa$$w0RD");
// Create client
ComputeManagementClient client = ComputeManagementService.create(config);
try {
    // Operation call
    VirtualMachineListResponse response = client.getVirtualMachineOperations().list("test-rg");
    List<VirtualMachine> vmList = new ArrayList<>();
    if (response != null) {
        vmList.addAll(response.getVirtualMachines());
    }
    while (response.getNextLink() != null) 
        // Get next page
        response = client.getVirtualMachineOperations().listNext(response.getNextLink());
        if (response != null) {
            vmList.addAll(response.getVirtualMachines());
        }
    }
} catch (ServiceException ex) {
    Logger.log("Error with status code " + ex.getHttpStatusCode() + " and message: " + ex.getMessage());
} catch (IOException | URISyntaxException ex) {
    Logger.log("Other exception thrown!");
}
```

### 1.0
```java
// Create client
ComputeManagementClient client = new ComputeManagementClientImpl(
        "localhost:3000",
        new UserTokenCredentials("username", "Pa$$w0RD"));
client.setSubscriptionId("subscription id in GUID");
try {
    // Operation call
    Page<VirtualMachine> response = client.getVirtualMachines().list("test-rg");
    List<VirtualMachine> vmList = new ArrayList<>();
    if (response != null) {
        vmList.addAll(response.getItems());
    while (response.getNextLink() != null) 
        // Get next page
        response = client.getVirtualMachines().listNext(response.getNextLink());
        if (response != null) {
            vmList.addAll(response.getItems());
        }
    }
} catch (CloudException ex) {
    Logger.log("Error with status code " + ex.getResponse().code() + " and message: " + ex.getMessage());
} catch (IllegalArgumentException ex) {
    Logger.log("Some parameters are invalid");
} catch (IOException ex) {
    Logger.log("Serialization exception usually");
}
```

**Comment by zfeng-cliqr** [comment]
Is there a plan to support deploying classic VMs (Microsoft.ClassicCompute/virtualMachines) using the Azure Resource Manager interface in 1.0 or later release? Since this is allowed through the UI and powershell, which I can deploy a classic vm to a resource group. I think it should be possible to do that in the SDK too.

**Comment by jianghaolu** [comment]
@zfeng-cliqr see my reply in https://github.com/Azure/azure-sdk-for-java/issues/573. Thanks!

**Comment by zfeng-cliqr** [comment]
@jianghaolu What's the targeted release date for sdk 1.0.0?

**Comment by jianghaolu** [comment]
@zfeng-cliqr 1.0.0-beta1 should be available by end of month. 1.0.0 needs a couple months' consolidation.

**Comment by grantfei** [comment]
How can I get the official build of java sdk? when will the sdk be released officially?
Thanks

**Comment by jianghaolu** [comment]
@grantfei See #637 & #643

**Comment by jianghaolu** [comment]
This has been implemented in 1.0.0-beta1. Closing.

beta2 will mostly focus on the developer experience that is experimented and prototyped here: https://github.com/microsoft/azure-shortcuts-for-java.
